design-proposals/golden-image-delivery-and-update-pipeline.md

```diff
@@ -0,0 +1,287 @@
+* Golden Disk Image Namespace: A namespace that specifically exists to host golden disk image PVCs. VMs can use CDI to cross namespace clone PVCs from this golden disk image namespace anywhere else in the cluster.
+
+## Goals
+
+* Pipeline capable of importing and keeping golden images up-to-date
+* Zero operational downtime during disk updates
+* No mutating existing PVCs during the import/update flow
+* Compatible with DataVolumes
+* Compatible with Templates (meaning helm and ocp templates)
+* Transitions seamlessly with PVC Volume Populators in the future
+
+## Non Goals
+
+* Not a pipeline for updating disks attached to existing VMs
+
+## User Stories
+
+* As a vendor/distributor of OS images, I would like the most fresh version of my OS to reach the KubeVirt clusters that want it.
+* As a user creating VMs, I want commonly used operating system disk images to automatically be made available to me without requiring uploading and managing my own os boot disk images.
+* As a user creating VMs from a template, I want to be able to stamp out a new VM from a template which will always use the latest operating system version available.
+* As a cluster admin managing VM images, I want the ability to automate sourcing in operating system disk image updates and make them available to users. This should come complete with monitoring and alerts to notify me if the automation fails.
+
+## Repos
+
+* **CDI**: Provides API (DataSource) to act as a mutable symbolic link to a PVC that can be referenced in a DataVolume. 
+* **CDI**: Provides API (DataImportCron) to handle recurring polling/importing disk images as PVCs into a golden image namespace using the CDI import flows
+* **SSP**: Manages setting up DataImportCrons objects with containerDisk image sources for commonly supported operating systems. Adds monitoring and alerting for pipeline failures.
+* **Common Templates**: Transitions to use CDI’s DataSource as DataVolumeTemplate source which ensures templates always reference the most up-to-date operating system disk images.
+* **HCO**: Pass in golden image container registries to ssp operator in order to setup automation import pipeline.
+* **OCP UI**: Needs to ensure DataSource is linked properly to a PVC in order to determine whether the boot image for a common template is available. 
+
+
+# Design
+
+![Golden Image Importer](golden-image-delivery-images/importer-design.png "Golden Image Importer Design")
+
+A **DataImportCron** object results in its corresponding controller polling for updates to a container image on a recurring interval. When updates to a container image are detected, the DataImportCron will automatically import the contents into a new PVC (using the standard CDI import flows) residing in a specific namespace with a unique name. The DataImportCron may need to implement a PVC garbage collection policy for removing outdated PVCs after an expiration period has been exceeded.
```

Comment:
Do we need an API on HCO to configure/tune the image garbage collector or golden image retention policy?

Reply:
To start with, i think it's reasonable for the HCO to not expose tunings related to image retention and stick with some sort of sane default behavor... something like garbage collecting all out dated images after a new disk is successfully imported. This will be something the hco can set internally on the ssp CR when defining the DataImportCrons

Maybe someday we'll need to expose retention polices on the HCO CR, but my opinion is that we can start out without it though.

Reply:
Tbh. I expected that HCO will reuse/import the API structs exposed by SSP for this.
Just plain passthrough of the relevant options. The rest should be handled by defaults on the schema.

Reply:
> Tbh. I expected that HCO will reuse/import the API structs exposed by SSP for this. Just plain passthrough of the relevant options. The rest should be handled by defaults on the schema.

I don't want to get into defining how advanced HCO CR tunings will work here. It's not necessary for the initial MVP of this feature. Lets get some experience under our belts with this feature and then have a future discussion about what (if any) advanced tunings should be exposed on the HCO CR. 

To start with I'd like to see the HCO expose a simple API that only allows opting into automated disk image deliver/updates. The HCO bundle will need to be built with an understanding of what disk sources are available for this automated disk import/update feature and those details will likely live within the CSV.

When a user toggles the opt in tuning on the HCO CR, the HCO then configures the changes necessary on the SSP operator.

@tiraboschi Does this sound like a reasonable objective to you for the first iteration of this feature?

Reply:
+1 from me on this

Reply:
I revised the `HCO: Integration with SSP Operator` header to reflect this.